### PR TITLE
[SPARK-32365][SQL] Add a boundary condition for negative index in regexp_extract

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
The current implement of regexp_extract will throws a unprocessed exception show below:
SELECT regexp_extract('1a 2b 14m', 'd+' -1)

```
java.lang.IndexOutOfBoundsException: No group -1
java.util.regex.Matcher.group(Matcher.java:538)
org.apache.spark.sql.catalyst.expressions.RegExpExtract.nullSafeEval(regexpExpressions.scala:455)
org.apache.spark.sql.catalyst.expressions.TernaryExpression.eval(Expression.scala:704)
org.apache.spark.sql.catalyst.optimizer.ConstantFolding$$anonfun$apply$1$$anonfun$applyOrElse$1.applyOrElse(expressions.scala:52)
org.apache.spark.sql.catalyst.optimizer.ConstantFolding$$anonfun$apply$1$$anonfun$applyOrElse$1.applyOrElse(expressions.scala:45)
```

### Why are the changes needed?
Fix a bug `java.lang.IndexOutOfBoundsException: No group -1`

### Does this PR introduce _any_ user-facing change?
Yes

### How was this patch tested?
new UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 1]
```diff
@@ -412,7 +412,10 @@ case class RegExpReplace(subject: Expression, regexp: Expression, rep: Expressio
 
 object RegExpExtract {
   def checkGroupIndex(groupCount: Int, groupIndex: Int): Unit = {
-    if (groupCount < groupIndex) {
+    if (groupIndex < 0) {
+      throw new IllegalArgumentException(
+        s"The specified group index cannot be less than zero")
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:417]: nit: remove `s` in the head.

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: OK. Thanks for your remind.

**Comment by maropu** [comment]
Nice catch, @beliefer . For better user's documents, could you add `arguments` in `ExpressionDescription` of `RegExpExtract` ? I think we could describe `index should be 0 or positive` there.

**Comment by beliefer** [comment]
@maropu OK.

**Comment by beliefer** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 2]
```diff
@@ -426,6 +428,14 @@ object RegExpExtract {
  */
 @ExpressionDescription(
   usage = "_FUNC_(str, regexp[, idx]) - Extracts a group that matches `regexp`.",
+  arguments = """
+    Arguments:
+      * str - a string expression.
+      * regexp - a string expression. a string representing a regular expression.
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:434]: Shall we remove the redundant `a string expression`?
```
- * regexp - a string expression. a string representing a regular expression.
+ * regexp - a string representing a regular expression.
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 3]
```diff
@@ -426,6 +428,14 @@ object RegExpExtract {
...
+          The regex string should be a Java regular expression.
+      * idx - a int expression that representing the group index. The group index should be
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:436]: `a int` -> `an integer`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 4]
```diff
@@ -426,6 +428,14 @@ object RegExpExtract {
...
+          0 or positive. If `idx` is not specified, the default group index value is 1.
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:437]: `or or positive` -> `non-negative`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 2 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:434]: OK.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 3 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:436]: Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 4 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:437]: OK.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @beliefer and @maropu .
Merged to master/3.0. (The last commit is only for comment-changes)

**Comment by beliefer** [comment]
@dongjoon-hyun @maropu Thanks for your help!

**Pull request closed without merging**